returnn/tf/util/data.py

```diff
@@ -427,41 +447,127 @@ def short_repr(self):
       """
       :rtype: str
       """
-      return "Var{%r}" % (self.dim_tag.description,)
+      return "Packed{%r}" % (self.dim_tag.description,)
 
-  def __init__(self, base=None, beam=None, virtual_dims=None):
+  def __init__(self, base, new_dim, new_dim_index=None):
     """
     :param BatchInfo|None base:
       If this is extended or based on another batch.
       Except of the batch dim of the dataset minibatch,
       we would always have a base.
-    :param SearchBeam|None beam:
-    :param list[BatchInfo.VirtualDimBase] virtual_dims:
+    :param BatchInfo.VirtualDimBase|None new_dim:
+    :param int|None new_dim_index:
 
     In most cases, this constructor would probably not be used directly by the user.
     """
     self.base = base
-    self.beam = beam
-    if not virtual_dims:
-      assert base and beam
-      virtual_dims = base.virtual_dims + [BatchInfo.BeamDim(beam=beam)]
-    self.virtual_dims = virtual_dims
+    virtual_dims = list(base.virtual_dims) if base else []
+    if new_dim:
+      if new_dim_index is None:
+        assert not virtual_dims
+        new_dim_index = 0
+      if new_dim_index < 0:
+        assert new_dim_index == -1
+        virtual_dims.append(new_dim)
+      else:
+        virtual_dims.insert(new_dim_index, new_dim)
+    self.virtual_dims = virtual_dims  # type: typing.List[BatchInfo.VirtualDimBase]
     self._dim = None  # type: typing.Optional[typing.Union[tf.Tensor,int]]
+    self._root_beam_dims_by_beam_name = {}  # type: typing.Dict[str,BatchInfo.BeamDim]
```

Comment:
Rename what? This attrib? Or `get_global_base()` (to `get_root_base()`?)? Any suggestion for a better name?

Reply:
Hm, not so sure. I think if the concept of a "global batch dim" is clear (see comment above), then "global" would be more descriptive - so I would rename the attribute. I'd also add to the docstring that it gives the BatchInfo with just the virtual global batch dim

Reply:
What about `_packed_dims_by_dim_tag`? Why is the logic there different (not global base?)?

Reply:
Well, with "root" I wanted to refer to the object you end up when you follow the bases (which is kind of a "parent") (so following filesystem terminology, going up the parents up to the root). That the root object is also the batch info with global batch dim is not relevant in this case.

That's also the difference to `_packed_dims_by_dim_tag`.

Those `._root_...` dicts are supposed to be used only from the root object, because the lookup should be global. E.g. the `BeamDim` object is globally the same, just depending on the beam. Normally I would have used class attributes, but I did not want to make them global for the whole process but just the current network.

Those non-root dicts are separate per object (batch info). A `PackedDim` will be different for the same `DimensionTag` on different batch info, because it depends on the batch info as well.

I agree, this is unclear. But I don't really know how to make it more clear. Maybe "global" instead of "root" is indeed better, but I think in the end it is just as confusing. I could add a longer comment before.

Reply:
I just realize, `make_global_broadcast_batch_info` is problematic, because it doesn't have the same root, and then also these dicts would not work...

Reply:
Ok, I renamed it now to `..global..`, and also added some more documentation.
(And about `make_global_broadcast_batch_info`, I added some extra checks.)

Reply:
Looks good how it is now, thanks! Also with the added documentation and the new comments.
Actually, it probably does not matter that much, it's only a private attribute and the intentions of the public methods of `BatchInfo` were already very clear before. But still, good!